docs/proposals/multi-tenant-networking.md

```diff
@@ -0,0 +1,202 @@
+
+**A new `network` resource will be added**
+
+```go
+// NetworkStatus is information about the current status of a Network.
+type NetworkStatus struct {
+    // Phase is the current lifecycle phase of the network.
+    Phase NetworkPhase `json:"phase,omitempty"`
+}
+
+type NetworkPhase string
+
+// These are the valid phases of a network.
+const (
+    // NetworkInitializing means the network is just accepted by system
+    NetworkInitializing NetworkPhase = "Initializing"
+    // NetworkActive means the network is available for use in the system
+    NetworkActive NetworkPhase = "Active"
+    // NetworkPending means the network is accepted by system, but it is still
+    // processing by network provider
+    NetworkPending NetworkPhase = "Pending"
+    // NetworkFailed means the network is not available
+    NetworkFailed NetworkPhase = "Failed"
+    // NetworkTerminating means the network is undergoing graceful termination
+    NetworkTerminating NetworkPhase = "Terminating"
+)
+
+// Subnet is a description of a subnet
+type Subnet struct {
+    // CIDR of this subnet
+    CIDR string `json:"cidr"`
+    // Gateway of this subnet
+    Gateway string `json:"gateway"`
+}
+
+// NetworkSpec is a description of a network
+type NetworkSpec struct {
+    // Which network provider/plugin provides this network and which will handle pod setup
+    Provider string `json:"provider,omitempty"`
+
+    // UUID specified by the network provider
+    Uuid string `json:"uuid,omitempty"`
+
+    // The network may contain zero or more subnets
+    Subnets map[string]Subnet `json:"subnets,omitempty"`
+}
+
+// Network describes an instantiated network
+type Network struct {
+    unversioned.TypeMeta `json:",inline"`
+
+    // Standard object's metadata.
+    // More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+    ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec defines the behavior of the Network.
+    Spec NetworkSpec `json:"spec,omitempty"`
+
+    // Status describes the current status of a Network
+    Status NetworkStatus `json:"status,omitempty"`
+}
+
+// NetworkList is a list of Networks
+type NetworkList struct {
+    unversioned.TypeMeta `json:",inline"`
+    // Standard list metadata.
+    // More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#types-kinds
+    unversioned.ListMeta `json:"metadata,omitempty"`
+
+    // Items is the list of Network objects in the list
+    Items []Network `json:"items"`
+}
+```
+
+**Network will be added to pod spec**
+
+```go
+// PodSpec is a description of a pod
+type PodSpec struct {
+  ...
+  // Networks the pod is assigned to
+  Networks []string `json:"networks,omitempty"`
+  ...
+}
+```
+
+## Examples
+
+### Create a network with overlapping 
+
+```yaml
+apiVersion: v1
+kind: Network
+metadata:
+  name: net1
+spec:
+  provider: mycorp/fancy-network-provider
+  uuid: 2a9aafcc-6041-4fb8-9006-1f6095102850
+  subnets:
```

Comment:
I think you should make the subnet to be an individual resource, like subnet in openstack. Subnet isn't determined or fixed when the network being creating. If we put subnet definition into network resource. Creating a new subnet must do through update network. It would be a little ugly.